File: trabalho/trab2/codigo/C/Trab2.c
/* De:Lucas Tatsuya Tanaka */
/* DRE:118058149 */
/* Email:tanaka@dcc.ufrj.br */
/* Github:https://github.com/tatsutanka */
/* Data:dd/mm/aa*/

#include<stdio.h>
#include<stdlib.h>
#include<limits.h>
#include<pthread.h>
#include<unistd.h>
#include"Queue.h"
#include"Threads.h"

//variaveis globais 
pthread_mutex_t mutex;// varivel de lock 
pthread_cond_t cond_leit,cond_escr; // variaveis de condicao 
int leit,escr;// indicadores de quantidade de leitores e escritores
int nLeitores; // numero de Threads Leitoras Criadas
int nEscritores; // numero de Threads Escritoras criadas

// main
int main(int argc, char * argv[]){
    pthread_t * tid; // indentificador de Threads
    Queue *q_key;// fila de compartilhada entre as Threads valores;
    Queue *q_id;// fila de compartilhada entre as Threads ids;
    Queue *q_leitura;// fila de compartilhada entre as Threads lidas;
    Estrutura *e; // Variavel compartilhada entre as Threads
    Monitor *m;

    //
    if(argc < 3){
        fprintf(stderr,"Digite: %s <Numero de sensores> <Numero de atuadores>",argv[0]);
        return 1;
    }
    nLeitores = atoi(argv[1]);
    nEscritores = atoi(argv[2]); 
    
    //------------------alocando espaco para as variaveis
    //aloca espaco para os indentificadores de Threads
    tid = (pthread_t *) malloc(sizeof(pthread_t) * (nLeitores + nEscritores));
    if(tid == NULL){fprintf(stderr,"ERRO --malloc"); return 3;}
    
    //aloca Queue
    q_key = createQueue(60);//capacidade de 60 definida
    q_id = createQueue(60);
    q_leitura = createQueue(60);

    //aloca o monitor 
    pthread_mutex_init(&mutex,NULL);
    pthread_cond_init(&cond_escr,NULL);
    pthread_cond_init(&cond_leit,NULL);

    m->cond_escr = &cond_escr;
    m->cond_leit = &cond_leit;
    m->escr = &escr;
    m->leit = &leit;
    m->mutex = &mutex;

    //aloca espaco para a Estrutura e
    e = (Estrutura*) malloc(sizeof(Estrutura) * (nLeitores + nEscritores));
    if(e == NULL){fprintf(stderr,"ERRO --malloc"); return 3;}
    //inicializa a Estrutura e
    for(int i = 0; i < (nLeitores+nEscritores); i++){
        e[i] = *createEstrutura(&e[i],q_key,q_id,q_leitura,m,i); 
    }

    //------------------criadno Threads
    //cria as threads leitoras
    for(int i=0; i<nLeitores; i++) {
      if(pthread_create(&tid[i], NULL, Leitor,&e[i])) exit(-1);
    } 
    
    //cria as threads escritoras
    for(int i=0; i<nEscritores; i++) {
      if(pthread_create(&tid[i+nLeitores], NULL, Escritor, &e[i+nLeitores])) exit(-1);
    } 

    //Monstradno os resultados

    //espera do termino das Threads
    return 0;
}

File: trabalho/trab2/codigo/Trab2.java
/* De:Lucas Tatsuya Tanaka */
/* DRE:118058149 */
/* Email:tanaka@dcc.ufrj.br */
/* Github:https://github.com/tatsutanka */
/* Data:11/05/2021*/


import java.lang.Math; // biblioteca responsavel por numeros aleatorios

// Estrutura compartilhada entre as Threads
class Estrutura{
    // intendificador do sensor
    //indentifcador de leitura = quantos leitores ja acessaram 
    // a variavel 

    public Estrutura(){

    }
}

//monitor que implementa a logica do padrao leitor/escritor
class LE{

    private int leit, escr;

    public LE(){
        this.leit = 0;// leitores lendo (0 ou mais)
        this.escr = 0;//escritor escrevendo (0 ou 1)
    }

    public synchronized void EntLeitor(int id) {
        try {
            while (this.escr > 0) {
                wait(); //bloqueia pela condicao logica
            }
            this.leit++; 
        } catch (InterruptedException e)         
    }

    public synchronized void SaiLeitor(int id ) {
        this.leit--;
        if(this.leit == 0){
            this.notify(); // libera escritor caso exista escritor bloqueado
        }
        
    }
    public synchronized void EntEscritor(int id) {
        try {
                while (this.leit > 0 || this.escr > 0) {
                    wait(); 
                }
                this.escr++;
             } catch (InterruptedException e) {
    }
    public synchronized void SaiEscitor(int id) {
       this.escr--; // registra a saida do escritor
       notifyAll(); //libera leitores e escritores ( caso existam bloqueados)
    }
}



//Aplicacao de exemplo--------------------------------------------------------
//Leitor = atuador 
// o atuador deve 
class Leitor extends Thread{
    int id;
    Estrutura e;//Estrutura compartilhada entre as Threads
    LE monitor;//objeto monitor para coordenacao da execucao 
    
    public Leitor(int id, LE m,Estrutura e) {
        this.e = e;
        this.id = id; 
        this.monitor = m;
    }

    public void run() {
       try {
            for(;;){
                this.monitor.EntLeitor();
                Thread.sleep(2000);
                // caso as ultimas 5 leituras derem acima de 
                // 35 graus emitir luz vermelho
                //
                // caso as ultimas 15 leituras derem acima de 
                // 35 graus emitir luz amarelo
                // 
                // 5 leitores indentifcados derem maio que 35
                // alerta amarelo
                //
                // caso nao de em nada o atuador vai dar um 
                // sinal de normal 
                this.monitor.SaiLeitor();
            }
        } catch (InterruptedException e) {return;}
    }

}
//--------------------------------------------------------
//Escritor = sensor
//Os sensores deverao medir a temperatura em intervalos
//de aproximadamente 1 segundo  o valor medio simulado etre 
//[25 ... 40] 
class Escritor extends Thread{
    int id;
    private simulado = 0;
    Estrutura e; //Estrutura compartilhada entre as Threads
    LE monitor;//objeto monitor para coordenacao da execucao 

    public Escritor(int id, LE m,Estrutura e) {
        this.e = e;
        this.id = id; 
        this.monitor = m;
    }

    public void run() {
       try {
            for(;;){
                this.monitor.EntEscritor();
                Thread.sleep(1000);
                simulado = ((int) (Math.random() * (40 - 25))) + 10);
                // sempre que o valor simulado for maior que 
                // 30 o valor vai ser escrito na variavel 
                // compartilhada com o valor do indentificador 
                // do sendsor e o indentificador de leitura 
                this.monitor.SaiEscitor();
            }
        } catch (InterruptedException e) {return;} 
    }
}

//--------------------------------------------------------
//classe principal 
class Trab2 {
    static final int L = 4; // numero de Threads Leitoras
    static final int E = 3; // numeor de Threads Escritoras

    public static void main(String[] args) {
        int i;
        Estrutura e = new Estrutura();
        LE monitor = new LE();
        Leitor[] l = new Leitor[L];
        Escritor e = new Escritor[E];
        
        
        if(args.length < 2){
         
        }
        // transformar String em numero 
        // L e E serao especificados na linha de comando 
        
        //inicia as Threads
        for(i=0; i<L; i++) {
           l[i] = new Leitor();
           l[i].start(); 
        }
        for(i=0; i<E; i++) {
           e[i] = new Escritor();
           e[i].start(); 
        }

        // Espera o terminao das Threads quado o programa é 
        // chamdo para o termino 

    }

}
